iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp

```diff
@@ -20,15 +22,57 @@ namespace iree_compiler {
 
 namespace {
 
-/// Pattern to combine instructions across ForOp boundary. It is common when
-/// doing incremental lowering to generate transient ops that cancel each others
-/// out. Canonicalization usually clean up those operations. When the value is
-/// loop carried, MLIR canonicalization currently doesn't remove the redundant
+/// Folds a value that has its definition and usage across SCF region
+/// boundaries if possible. Returns null value otherwise.
+///
+/// Such cases won't have def-use relationship via direct SSA reference; rather
+/// it's via value yielding and usage of the SCF op as a whole.
+static Value foldValueDefUse(Operation* user, Operation* def) {
+  if (auto shapeCast = dyn_cast_or_null<vector::ShapeCastOp>(user)) {
+    if (auto souceOp = dyn_cast_or_null<vector::ShapeCastOp>(def)) {
+      if (shapeCast.getType() == souceOp.source().getType())
+        return souceOp.source();
+    }
+  } else if (auto extractOp = dyn_cast_or_null<vector::ExtractOp>(user)) {
+    if (auto broadcastOp = dyn_cast_or_null<vector::BroadcastOp>(def)) {
+      if (extractOp.getType() == broadcastOp.getSourceType())
+        return broadcastOp.source();
+      // If we are extracting a scalar element, it's also okay to fold.
+      if (extractOp.getType().isIntOrFloat()) return broadcastOp.source();
+    }
+  } else if (auto targetOp =
+                 dyn_cast_or_null<UnrealizedConversionCastOp>(user)) {
+    if (auto sourceOp = dyn_cast_or_null<UnrealizedConversionCastOp>(def)) {
+      if (sourceOp->getNumOperands() == 1 && targetOp->getNumResults() == 1 &&
+          sourceOp->getOperandTypes().front() ==
+              targetOp.getResultTypes().front())
+        return sourceOp.inputs().front();
+    }
+  }
+  return Value();
+}
+
+/// Moves all ops from the `source` block into the `dest` block and updates the
+/// yield terminator op to yield values in `results`.
+static void transferBody(Block* source, Block* dest, ArrayRef<Value> results,
+                         RewriterBase& rewriter) {
+  // Move all operations to the destination block.
+  rewriter.mergeBlocks(source, dest, dest->getArguments());
+  // Replace the yield op by one that returns only the used values.
+  auto yieldOp = cast<scf::YieldOp>(dest->getTerminator());
+  yieldOp.getOperation()->setOperands(results);
+}
+
+/// Fold reciprocal ops across ForOp boundary. It is common when doing
```

Comment:
The name reciprocal sounds a bit odd to me in this context (makes me think of 1/x kind of reciprocal). Should. we say canceling ops or maybe another name? (If any native speakers are around maybe they have suggestions).
Not a blocker though.

Reply:
Yeah I don't have a good term here too.. Changed.